File: app/src/test/java/net/squanchy/search/engines/SpeakerSearchEngineTest.java
package net.squanchy.search.engines;

import net.squanchy.speaker.domain.view.Speaker;

import org.junit.Test;

import static net.squanchy.speaker.domain.view.SpeakerFixtures.aSpeaker;
import static org.fest.assertions.api.Assertions.assertThat;

public class SpeakerSearchEngineTest {

    private static final Speaker ANY_SPEAKER = aSpeaker().build();

    private static final Speaker MATCHING_SPEAKER = aSpeaker()
            .withName("Banana Joe")
            .build();

    private final SpeakerSearchEngine searchEngine = new SpeakerSearchEngine();

    @Test
    public void givenEmptyQuery_whenMatching_thenReturnsFalse() {
        String query = "";

        boolean matches = searchEngine.matches(ANY_SPEAKER, query);

        assertThat(matches).isFalse();
    }

    @Test
    public void givenQueryShorterThanTwoCharacters_whenMatching_thenReturnsFalse() {
        String query = "a";

        boolean matches = searchEngine.matches(MATCHING_SPEAKER, query);

        assertThat(matches).isFalse();
    }

    @Test
    public void givenOfAtLeastTwoCharacters_whenMatching_thenReturnsTrue() {
        String query = "na";

        boolean matches = searchEngine.matches(MATCHING_SPEAKER, query);

        assertThat(matches).isTrue();
    }

    @Test
    public void givenSpeakerWithNameMatchingQueryExactly_whenMatching_thenReturnsTrue() {

        boolean matches = searchEngine.matches(MATCHING_SPEAKER, "Banana Joe");

        assertThat(matches).isTrue();
    }

    @Test
    public void givenSpeakerWithNameContainingQuery_whenMatching_thenReturnsTrue() {

        boolean matches = searchEngine.matches(MATCHING_SPEAKER, "anana J");

        assertThat(matches).isTrue();
    }

    @Test
    public void givenSpeakerWithNameMatching_butWithDifferentCasing_whenMatching_thenReturnsTrue() {

        boolean matches = searchEngine.matches(MATCHING_SPEAKER, "anana J");

        assertThat(matches).isTrue();
    }

    @Test
    public void givenSpeakerWithNameMatchingQueryExactly_butWithDiacritics_whenMatching_thenReturnsTrue() {

        boolean matches = searchEngine.matches(MATCHING_SPEAKER, "anÀna J");

        assertThat(matches).isTrue();
    }
}

File: app/src/test/java/net/squanchy/search/engines/SpeakerSearchEngineTest.kt
package net.squanchy.search.engines

import com.google.common.truth.Truth.assertThat
import net.squanchy.speaker.domain.view.aSpeaker
import org.junit.Test

class SpeakerSearchEngineTest {

    private val searchEngine = SpeakerSearchEngine()

    @Test
    fun givenEmptyQuery_whenMatching_thenReturnsFalse() {
        val query = ""

        val matches = searchEngine.matches(ANY_SPEAKER, query)

        assertThat(matches).isFalse()
    }

    @Test
    fun givenQueryShorterThanTwoCharacters_whenMatching_thenReturnsFalse() {
        val query = "a"

        val matches = searchEngine.matches(MATCHING_SPEAKER, query)

        assertThat(matches).isFalse()
    }

    @Test
    fun givenOfAtLeastTwoCharacters_whenMatching_thenReturnsTrue() {
        val query = "na"

        val matches = searchEngine.matches(MATCHING_SPEAKER, query)

        assertThat(matches).isTrue()
    }

    @Test
    fun givenSpeakerWithNameMatchingQueryExactly_whenMatching_thenReturnsTrue() {

        val matches = searchEngine.matches(MATCHING_SPEAKER, "Banana Joe")

        assertThat(matches).isTrue()
    }

    @Test
    fun givenSpeakerWithNameContainingQuery_whenMatching_thenReturnsTrue() {

        val matches = searchEngine.matches(MATCHING_SPEAKER, "anana J")

        assertThat(matches).isTrue()
    }

    @Test
    fun givenSpeakerWithNameMatching_butWithDifferentCasing_whenMatching_thenReturnsTrue() {

        val matches = searchEngine.matches(MATCHING_SPEAKER, "anana J")

        assertThat(matches).isTrue()
    }

    @Test
    fun givenSpeakerWithNameMatchingQueryExactly_butWithDiacritics_whenMatching_thenReturnsTrue() {

        val matches = searchEngine.matches(MATCHING_SPEAKER, "anÀna J")

        assertThat(matches).isTrue()
    }

    companion object {

        private val ANY_SPEAKER = aSpeaker()

        private val MATCHING_SPEAKER = aSpeaker(name = "Banana Joe")
    }
}
